### ISSUE TITLE
Internet Explorer IndexSizeError on Drawer.blendSketch context.drawImage

### ISSUE THREAD

**Issue by adgoncal** [openseadragon/openseadragon#1033]
Demo: http://plnkr.co/edit/ormkIn

It works fine on Chrome, but IE throws an error and the image freezes (cannot pan or zoom).
There are multiple Google results for "ie drawImage IndexSizeError" but I could not get it to work.

IE spec says:

> **IndexSizeError** The numeric arguments are not valid (for example, the destination is a 0x0 rectangle). Versions earlier than Internet Explorer 10 use INDEX_SIZE_ERR.

https://msdn.microsoft.com/en-us/library/ff975414(v=vs.85).aspx

EDIT:
`homeFillsViewer: true` makes IE throw the error consistently.
`homeFillsViewer: false` still throws the error, but if you pan or zoom when the image first shows up, sometimes it does not throw the error and everything works afterwards.

https://github.com/openseadragon/openseadragon/blob/v2.2.1/src/drawer.js#L426

**Comment by avandecreme** [comment]
What version of IE is it? Does it happen on Edge as well?
I do not have access to a windows machine to reproduce.
To debug that further, I would display the sketch canvas and see what is in there when the crash occurs. You can do that by attaching the sketchCanvas to the DOM in drawer.js.
Any chance you could try that?

**Comment by adgoncal** [comment]
Tested browsers:

Works:

- Chrome Version 53.0.2785.116 m
- Firefox 47.0

Does not work:

- Internet Explorer 11
- Internet Explorer 10
- Microsoft Edge 14

I will try attaching the sketchCanvas to the DOM and will post again with results.

**Comment by adgoncal** [comment]
```javaScript
var originalBlendSketch = OpenSeadragon.Drawer.prototype.blendSketch;
OpenSeadragon.Drawer.prototype.blendSketch = function() {
  try {
    originalBlendSketch.apply(this, arguments);
  } catch(err) {
    if(this.sketchCanvas) {
      $(this.sketchCanvas).css({
        position: 'absolute',
        bottom: '0',
        right: '0',
        zIndex: '9999',
      }).appendTo('body')
    }
    throw err;
  }
};

```

sketchCanvas looks fine. It is displaying the second tiledImage from tileSources array.

![image](https://cloud.githubusercontent.com/assets/10856791/18727746/290a7a68-8018-11e6-9a6b-50b8d6fe6d70.png)

**Comment by avandecreme** [comment]
Can you spot anything weird in the parameters passed to context.drawimage?

**Comment by adgoncal** [comment]
```JavaScript
sketchCanvas.width: 200
sketchCanvas.height: 200
bounds: {
  x: -34,
  y: 0,
  width: 268,
  height: 200
}
```

I noticed there is no error if I set:
```JavaScript
bounds.width = this.sketchCanvas.width;
bounds.height = this.sketchCanvas.height;
bounds.x = 0;
bounds.y = 0;
```

However, this draws the image in the wrong place.

**Comment by adgoncal** [comment]
BTW, the issue does not happen on this example using Deep-Zoom tiles instead of legacy-image-pyramid:
https://codepen.io/iangilman/pen/YXRLea

**Comment by adgoncal** [comment]
The same issue happens with a different `tileSource`:
http://plnkr.co/edit/bQeg30?p=preview

This time it only happens after you press toggle. 

I noticed that the issue only occurs when the image is not  entirely visible, either because of panning it outside the visible area or zooming in so the image is bigger than the visible area.

**Comment by avandecreme** [comment]
So maybe IE doesn't like the negative x?
Could you try avoiding that by adding those 2 conditions in blendSketch:
```javascript
            if (bounds.x < 0) {
                bounds.width += bounds.x;
                bounds.x = 0;
            }
            if (bounds.y < 0) {
                bounds.height += bounds.y;
                bounds.y = 0;
            }
            this.context.drawImage(
                this.sketchCanvas,
                bounds.x,
                bounds.y,
                bounds.width,
                bounds.height,
                bounds.x,
                bounds.y,
                bounds.width,
                bounds.height
            );
```

(I did not test it, so I might have forgotten something)

**Comment by adgoncal** [comment]
I tried that and got the same error.
```javaScript
sketchCanvas.width: 200
sketchCanvas.height: 200
bounds: {
  x: 0,
  y: 0,
  width: 234,
  height: 200
}
```

**Comment by avandecreme** [comment]
Ha maybe it doesn't like that `bounds.width` is above `sketchCanvas.width` or `canvas.width` either?
Could you enforce that as well?

**Comment by adgoncal** [comment]
BINGO!
```javaScript
var originalBlendSketch = OpenSeadragon.Drawer.prototype.blendSketch;
OpenSeadragon.Drawer.prototype.blendSketch = function(options) {
  if(this.useCanvas && this.sketchCanvas && options && options.bounds) {
    if(options.bounds.x < 0) {
      options.bounds.width += options.bounds.x;
      options.bounds.x = 0;
    }
    if(options.bounds.width > this.sketchCanvas.width) {
      options.bounds.width = this.sketchCanvas.width;
    }
    if(options.bounds.y < 0) {
      options.bounds.height += options.bounds.y;
      options.bounds.y = 0;
    }
    if(options.bounds.height > this.sketchCanvas.height) {
      options.bounds.height = this.sketchCanvas.height;
    }
  }
  try {
    originalBlendSketch.apply(this, arguments);
  } catch(err) {
    console.log(options.bounds.x, options.bounds.y, options.bounds.width, options.bounds.height)
    throw err;
  }
};
```

**Comment by avandecreme** [comment]
Excellent!
Would you be up to make a PR for this?

**Comment by adgoncal** [comment]
Will do, thanks for the help!

**Comment by avandecreme** [comment]
Excellent! I will leave this issue open until the fix is merged.

**Comment by iangilman** [comment]
Fixed by #1035
